### ISSUE TITLE
story deploy occasionally spits out a nasty OSX error

### ISSUE THREAD

**Issue by williammartin** [storyscript/cli#280]
```
$ story deploy
Compiling Stories…
Deploying app electronic-wormhole-8121... ⠋objc[26190]: +[__NSPlaceholderDate initialize] may have been in progress in another thread when fork() was called.
objc[26190]: +[__NSPlaceholderDate initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.
⠹
✔ Version 100 of your app has been queued for deployment.

Waiting for deployment to complete…  ⠹
✔ Deployment successful!
If your Story responds to HTTP requests, please visit:
  https://electronic-wormhole-8121.storyscriptapp.com/
```

**Comment by arturtamborski** [comment]
After a quick run down in the search results of for that error I can say that this is in fact a bug way up in the chain of libs, because it's in Python implementation itself - https://bugs.python.org/issue33725.
As always, there's a quick and dirty fix
```
export OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES
```
though I don't have a Mac so I can't confirm if that fixes anything.

Either way there's not much to be done other than waiting for upstream bug fix.

**Comment by williammartin** [comment]
Thanks for doing the legwork and finding this python issue.

It seems like there are two ways to tackle this:
 - Wait for Python 3.8 which changes the semantics of creating a new process from `fork/exec` to `spawn`
- Find out where in the CLI we are playing badly with multiprocessing and fix that specifically (but this could be down in some library)

As for `OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES` that seems to be papering over the issue. From what I can tell in `10.13` a warning message was added, and in `10.14` the process is forced to crash. With that in mind, disabling the crash would open up undefined behaviour - this may not be a problem depending on our usage, but since we don't actually know where the error is coming from, that's not a bet I would like to take!

**Comment by williammartin** [comment]
I think the issue here is in the sentry tracking:

https://github.com/storyscript/cli/blob/master/story/cli.py#L75-L108

For some reason, on Linux and Darwin, we are `fork`ing the request into another process. However, doing nearly anything between `fork` and `exec` is not supported behaviour. In this case we don't even `exec`, simply do the sentry request.

I think if we remove the forking behaviour here (though I'm a bit concerned about Chesterton's Fence here so need to speak to @judepereira) we'll prevent this error.

Separately, this code is also related to the broken test behaviour in https://github.com/storyscript/cli/pull/304

**Comment by williammartin** [comment]
Here's some further reading: https://wefearchange.org/2018/11/forkmacos.rst.html

This explicitly calls out the `requests` library as a concern:

```
The problem is that your program may implicitly call a +initialize method without you knowing it. In Python 3 for example, if you call into the popular requests library, this will end up calling into the _scproxy module to get the system proxies, and this will end up calling a +initialize method. So you better not use requests between fork() and exec*()! Setting the environment variable no_proxy = '*' will prevent this, and avoid the crash, but it also bypasses the system proxies, which is probably not what you want.
```

**Comment by williammartin** [comment]
I spoke to @judepereira and this code exists to not block users on the tracking, instead putting it in the background. We'll need to find another solution to this problem, but I guess we can't just remove the forking code.

**Comment by jayvdb** [comment]
Yes we can and should remove the forking code, but keep the user requirement of non-blocking.

The high level APIs of [threading](https://docs.python.org/3/library/threading.html) or [multiprocessing](https://docs.python.org/3/library/multiprocessing.html) could be used instead.  Sentry is already doing threading, so that would be the better choice.

But preferrably we go with a lighter approach of an async http library like [httpx](https://github.com/encode/httpx/async) (c.f. https://github.com/encode/requests-async which has been rolled into that library - and is way beyond alpha)

Lots of similar libraries exist: curequests, asks, trip.

Given how small this workload is, and there is no waiting on the result, it should be easy to get it working with any of these libraries.  Ideally we find one that is fairly low overhead in terms of processing and dependencies, and probably low complexity also so that writing functional tests isnt too difficult.

**Comment by williammartin** [comment]
I'm not sure that any "new process" solution will work while our CLI integration tests are written in a way that exposes https://github.com/storyscript/cli/issues/288#issuecomment-539418254

**Comment by jayvdb** [comment]
Ya, obviously need to get that fixed before merging anything significant.

**Comment by williammartin** [comment]
Just to be clear, I'm talking about the fact that the CLI test runner captures `SystemExit` exceptions, which results in multiple `pytest` processes (so new processes are out under that mechanism), not just the fact that the build is red (and yes I'd prefer we didn't merge anything while that is true also).

**Comment by jayvdb** [comment]
New processes are the least desirable of the suggested approaches, so no worries there.

Just noting, `ensure.py` is also doing blocking a http request which would benefit from being made async, but it isnt as simple a conversion as the tracking post, and it uses a set of servers which wont be re-used during the CLI invocation so the shared session management isnt so important.  Point is that there will be other uses of an async http library.

**Comment by judepereira** [comment]
Okay - let's not fork. Another approach which respects the requirement is:
1. If curl is available, background a curl call to the endpoint - however, the known issue here is that large payloads might not work reliably, due to a shell limitation. However, this is not a problem for us, since our payloads are small
2. If curl is unavailable, sacrifice the behaviour tracking code

**Comment by judepereira** [comment]
There's no requirement of an async http library yet, so let's not introduce one.

**Comment by judepereira** [comment]
PS - about the curl based solution - we can also wrap our own little Python script for this tracking which uses requests, and we execute this as a script in the background, thereby not calling fork altogether. This will also be cross platform, and not depend on curl.

**Comment by williammartin** [comment]
> There's no requirement of an async http library yet, so let's not introduce one.

This isn't really a reason to not introduce the library. There's no requirement to choose any particular implementation. Can you rephrase this in terms of the advantages and disadvantages of each approach?

**Comment by judepereira** [comment]
An async http library is not going to solve the original problem - the original problem is that the CLI was waiting for the behaviour tracking HTTP call to complete, when it really should've just exited. For that reason, we forked the behaviour tracking call into the background. 

Introducing an async http library is not going to solve this, since the main CLI process, will still need to wait for that HTTP call to complete safely, and then exit.

The background approaches of calling another executable should work since it satisfies the following requirements:
1. Don't call fork
2. Don't block the user for user behaviour HTTP tracking calls

**Comment by williammartin** [comment]
> An async http library is not going to solve the original problem - the original problem is that the CLI was waiting for the behaviour tracking HTTP call to complete, when it really should've just exited.

Gotcha, thanks for the explanation. I hadn't fully grokked this wasn't just about executing asynchronously in a typical long running or concurrent process scenario, but actually about not blocking the main CLI process from exiting.

**Comment by jayvdb** [comment]
An async http request does not need to block the main cli process.  It just runs on the side in the event loop, optionally in a separate thread depending on the async implementation in use.  The main process can check it if it wants to, but currently the code is ignoring any exceptions which occur in the tracking, so that hasnt been a part of the implementation.

**Comment by judepereira** [comment]
> It just runs on the side in the event loop, optionally in a separate thread depending on the async implementation in use

Consider the following two scenarios:
1. Running in a separate thread - unless I'm mistaken, the CLI main process will only exit when all threads exit
2. Running in the event loop, at the side - if the CLI process terminates early, the event loop does not get a chance to schedule and run this task. However, if we wait for it to schedule and execute the task, then we run into the original problem again - which is, the CLI waits for the behaviour tracking call to complete.

**Comment by jayvdb** [comment]
ya, I added a bit wrt multiprocessing if we *really* need to have tracking for CLI usage where the CLI invocation time was shorter than the time to send off the tracking json -- currently they are very small.  I would need to check, but I recall multiprocessing being able to handle subprocesses persisting after the main process has completed.  But that is an implementation detail.

This feels like unclear functional requirements, because in parts of the discussion the tracking can be completely non-functional for one type of implementation, but then it is considered mandatory that it completes even when the user has run something so small that the CLI couldnt even send a tiny packet, eliminating other alternatives.

Also the current behaviour on Windows is completely different, lacking the features stated as required, yet there is awareness that our userbase will include a large percentage of Windows users.

**Comment by judepereira** [comment]
True. Would you say that running a background process which would be a Python script be a viable approach? It would work on Windows, Linux, and Mac.

**Comment by jayvdb** [comment]
Obviously it is more viable compared to the existing code.

However that implementation is based on eliminating other options due to the decision that tracking minuscule usage of the CLI is so important that it should include creating background processes, which are not zero cost (time or resources), and having them lingering around unmanaged is not ideal.  The functional requirement of tracking users at all cost is the problem clouding the decision making here.

A lower resource and better technical solution is to use an async event loop, and kill the tracking if the CLI didnt do anything and needs to quit early, or pause for it to finish if something important was done by the user: something worth tracking.  Real users are going to frown on the excessive tracking, and being open source they will come and submit code to remove any unnecessary tracking.

**Comment by judepereira** [comment]
We're tracking things selectively only. Not every command has this, only commands that result in app lifecycle. However, tracking is made on a "best effort" approach. If the background process fails, it's acceptable. 

The premise is that every network call takes a few seconds on a poor connection, and hence we should only make network calls that are truly required during the lifecycle of the main CLI process.
